### PR TITLE
Time function improvements (fix #1912)

### LINKED ISSUE #1912: Wrong timezone names in strftime/strflocaltime for DST
**Describe the bug**
jq  does not print the timezone correctly expanding the %Z in the time format for strftime/strflocaltime when it comes to DST.
**To Reproduce**
Using the example from the wiki...
```
$ TZ=Europe/Rome ./jq-linux64 -cn 'now|strflocaltime("%Y-%m-%dT%H:%M:%S %Z")'
"2019-05-28T17:10:54 CET"

$ TZ=Europe/Rome date +"%Y-%m-%dT%H:%M:%S %Z"
2019-05-28T17:10:54 CEST
```
`date` reports, correctly, `CEST` as the timezone taking DST into account. `jq` does not.

**Expected behavior**
The output of `jq` to specify the correct timezone following DST

**Environment (please complete the following information):**
 - Linux Ubuntu 16.04
 - jq version 1.6

### PR DESCRIPTION
**Pull request by nicowilliams**
- Fixes #1912, and fixes / documents #2865.
- Makes `mktime` preserve any fractional seconds part.
- Adds `timelocal` and `timegm`.

### CODE REVIEW

**Comment by nicowilliams** [comment]
image

Gotta normalize `tm_zone` for tests, or maybe change the tests to normalize it.  Probably the latter.

**Comment by nicowilliams** [comment]
~For Windows we should try using `tzname` in `f_localtime()`.~

**Comment by nicowilliams** [comment]
We can also work around the lack of a `tm_gmtoff` field in some cases by checking the local and gm time and computing the difference, just as we can get the TZ name from `tzname`.

**Comment by nicowilliams** [comment]
> We can also work around the lack of a `tm_gmtoff` field in some cases by checking the local and gm time and computing the difference, just as we can get the TZ name from `tzname`.

Er, this is fraught.  For example, if I try to use `localtime_r()` and `gmtime_r()` then `difftime()` with a `tm` value that is in DST, I get the wrong result.  I'm inclined not to do this.

File: src/builtin.c [hunk 1]
```diff
@@ -1509,25 +1515,87 @@ static int jv2tm(jv a, struct tm *tm) {
...
+    return jv_invalid_with_msg(jv_string("timegm requires parsed datetime inputs"));
+#ifdef HAVE_TIMEGM
+  time_t t = timegm(&tm);
+  free(freeme);
+#elif defined(HAVE__MKGMTIME)
+  time_t t = _mkgmtime(&tm); // Windows has _mkgmtime()
+  free(freeme);
+#else
+  free(freeme);
+  return jv_invalid_with_msg(jv_string("tmigm is not supported"));
```
Comment (emanuele6) [inline src/builtin.c]: tmigm => timegm

**Comment by nicowilliams** [comment]
Now with some boundary tests (which found bugs).

**Comment by nicowilliams** [comment]
So `tm_zone` is `const char *` on Linux but `char *` or OS X.

File: src/builtin.c [hunk 2]
```diff
@@ -1446,65 +1446,108 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
...
-      }                                         \
-      t = jv_number_value(n);                   \
-      jv_free(n);                               \
-    } while (0)
+/*
+ * Utility function for jv2tm() that returns `ifnan` if `v` is not a number in
+ * any way, and clamps the value range of `v` to `imin` and `imax`.
+ */
+static int toint(jv v, int imin, int imax, int ifnan) {
+  if (jv_get_kind(v) != (JV_KIND_NUMBER) || jvp_number_is_nan(v)) {
```
Comment (nicowilliams) [inline src/builtin.c:1454]: So NaNs and also non-number `jv` values are treated as NaNs here :)

File: src/builtin.c [hunk 3]
```diff
@@ -1446,65 +1446,108 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
...
+  return (int)n;
+}
 
 static int jv2tm(jv a, struct tm *tm, double *frac, char **freeme) {
   int ret = 1;
 
   *freeme = NULL;
   memset(tm, 0, sizeof(*tm));
-  TO_TM_FIELD(tm->tm_year, a, 0);
+  if (   (tm->tm_year = toint(jv_array_get(jv_copy(a), 0),
```
Comment (nicowilliams) [inline src/builtin.c:1477]: ISO 8601 has an option for 5-digit years, but we'll go up to just 4, which means up to `9999`.  For now we won't support time values before 1970 (though we should).

File: src/builtin.c [hunk 4]
```diff
@@ -1446,65 +1446,108 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
...
+                              1969, 10000, 1969))                       < 1970
+      ||  tm->tm_year                                                   > 9999
+      || (tm->tm_mon  = toint(jv_array_get(jv_copy(a), 1), -1, 12,  -1)) < 0
```
Comment (nicowilliams) [inline src/builtin.c:1480]: Recall that
```
           struct tm {
               int tm_sec;    /* Seconds (0-60) */
               int tm_min;    /* Minutes (0-59) */
               int tm_hour;   /* Hours (0-23) */
               int tm_mday;   /* Day of the month (1-31) */
               int tm_mon;    /* Month (0-11) */
               int tm_year;   /* Year - 1900 */
               int tm_wday;   /* Day of the week (0-6, Sunday = 0) */
               int tm_yday;   /* Day in the year (0-365, 1 Jan = 0) */
               int tm_isdst;  /* Daylight saving time */
           };
```
So `tm_mon` must be between 0 and 11, inclusive, so we call `toint()` with `-1` and `12` instead and use `< 0` and `> 11` to check for invalid values.

File: src/builtin.c [hunk 5]
```diff
@@ -1446,65 +1446,108 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
...
+      ||  tm->tm_mon                                                    > 11
+      || (tm->tm_mday = toint(jv_array_get(jv_copy(a), 2),  0, 32,   0)) < 1
```
Comment (nicowilliams) [inline src/builtin.c:1482]: `int tm_mday;   /* Day of the month (1-31) */` which means we call `toint()` with `0` and `32` and check `< 1` and `> 31`.

File: src/builtin.c [hunk 6]
```diff
@@ -1446,65 +1446,108 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
...
+      ||  tm->tm_mday                                                   > 31
+      || (tm->tm_hour = toint(jv_array_get(jv_copy(a), 3), -1, 24,  -1)) < 0
```
Comment (nicowilliams) [inline src/builtin.c:1484]: `int tm_hour;   /* Hours (0-23) */`m so `-1`..`24`.

File: src/builtin.c [hunk 7]
```diff
@@ -1446,65 +1446,108 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
...
+      ||  tm->tm_hour                                                   > 23
+      || (tm->tm_min  = toint(jv_array_get(jv_copy(a), 4), -1, 60,  -1)) < 0
```
Comment (nicowilliams) [inline src/builtin.c:1486]: `int tm_min;    /* Minutes (0-59) */`, so `-1`..`60`.

File: src/builtin.c [hunk 8]
```diff
@@ -1446,65 +1446,108 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
...
+      ||  tm->tm_min                                                    > 59
+      || (tm->tm_sec  = toint(jv_array_get(jv_copy(a), 5), -1, 61,  -1)) < 0
```
Comment (nicowilliams) [inline src/builtin.c:1488]: `int tm_sec;    /* Seconds (0-60) */` (because of leap seconds), so `-1`..`61`.

File: src/builtin.c [hunk 9]
```diff
@@ -1446,65 +1446,108 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
...
+      ||  tm->tm_sec                                                    > 60
+      || (tm->tm_wday = toint(jv_array_get(jv_copy(a), 6), -1, 7,   -1)) < 0
```
Comment (nicowilliams) [inline src/builtin.c:1490]: `int tm_wday;   /* Day of the week (0-6, Sunday = 0) */`, so `-1`..`7`.

File: src/builtin.c [hunk 10]
```diff
@@ -1446,65 +1446,108 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
...
+      ||  tm->tm_wday                                                   > 6
+      || (tm->tm_yday = toint(jv_array_get(jv_copy(a), 7), -1, 366, -1)) < 0
```
Comment (nicowilliams) [inline src/builtin.c:1492]: `int tm_yday;   /* Day in the year (0-365, 1 Jan = 0) */`, so `-1`..`366`.

File: src/builtin.c [hunk 11]
```diff
@@ -1446,65 +1446,108 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
...
   }
   jv_free(v);
-#ifdef HAVE_TM_TM_GMTOFF
   v = jv_array_get(jv_copy(a), 9);
-  if (jv_get_kind(v) == JV_KIND_NUMBER)
-    tm->tm_gmtoff = jv_number_value(v);
-  else if (jv_is_valid(v))
-    ret = 0;
-  jv_free(v);
+  int gmtoff;
```
Comment (nicowilliams) [inline src/builtin.c:1516]: We validate `gmtoff` even if the local system doesn't have `tm_gmtoff`.

File: src/builtin.c [hunk 12]
```diff
@@ -1446,65 +1446,108 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
...
 #endif
-#ifdef HAVE_STRUCT_TM_TM_ZONE
+    break;
+  }
+  jv_free(v);
   v = jv_array_get(jv_copy(a), 10);
-  if (jv_get_kind(v) == JV_KIND_STRING) {
-    tm->tm_zone = *freeme = strdup(jv_string_value(v));
-    if (tm->tm_zone == NULL)
+  switch (jv_get_kind(v)) {
```
Comment (nicowilliams) [inline src/builtin.c:1531]: We validate the timezone name even if the local system does not have `tm_zone`.

File: tests/jq.test [hunk 13]
```diff
@@ -1580,6 +1580,15 @@ try mktime catch .
 ["a",1,2,3,4,5,6,7]
 "mktime requires parsed datetime inputs"
 
+# #2863
+[.[]|try (timegm | error) catch .]|[length,unique[]]
+[[1969],[10001],[1970,-1],[1970,12],[1970,11,0],[1970,11,32],[1970,11,30,-1],[1970,11,30,24],[1970,11,30,23,-1],[1970,11,31,23,60],[1970,11,30,23,59,-1],[1970,11,30,23,59,61],[1970,11,30,23,59,60,-1],[1970,11,30,23,59,60,7],[1970,11,30,23,59,60,4],[1970,11,30,23,59,60,4,-1],[1970,11,30,23,59,60,4,366],[1970,11,30,23,59,60,4,364,true,-43201],[1970,11,30,23,59,60,4,364,true,43201],[1970,11,30,23,59,60,4,364,true,0,true]]
```
Comment (nicowilliams) [inline tests/jq.test]: This tests boundary conditions.  So since we disallow years before 1969 (maybe we should allow a larger range), `[1969,...]` should fail immediately, so we have `[1969]` to test for that.

**Comment by nicowilliams** [comment]
I've left resolved comments to explain some things, mainly the range validation checks for all the fields of broken down time (which were wrong in an earlier version).

**Comment by nicowilliams** [comment]
Well, I think we should merge this in time for 1.7, but I'm clearing the milestone anyways.

**Comment by nicowilliams** [comment]
Ok, we should allow datetimes before 1970, so I'm making the range `1900..9999`, though perhaps on 64-bit systems we can go to `-9999..9999`?

EDIT: Yes, so let's go to `0001..9999` as the range for years.  It works on Linux.

**Comment by trantor** [comment]
mmm was this not merged before 1.7 ? @nicowilliams

**Comment by nicowilliams** [comment]
> mmm was this not merged before 1.7 ? @nicowilliams

@itchyny thought it was too risky for 1.7.  I intend to do a new release soon with this and a few other things.

**Comment by tianon** [comment]
Was feeling like I was going crazy with the off-by-one-hour behavior of `fromdate` and then found this -- any chance this gets picked back up? :innocent: :heart:

```console
$ jq -n '"2025-08-01T00:00:00Z" | fromdate | todate'
"2025-08-01T01:00:00Z"

$ TZ=UTC jq -n '"2025-08-01T00:00:00Z" | fromdate | todate'
"2025-08-01T00:00:00Z"
```

**Comment by itchyny** [comment]
@tianon The bug is likely fixed in jq 1.8.0. What version are you using?

**Comment by tianon** [comment]
Doh, I'm on 1.6 -- I see it's even fixed in the 1.7 in Debian Trixie.  Sorry for the noise!
